File: spline/spline.py
# -*- coding:utf-8 -*-


def drange(start, stop, step):
    r = start
    while r < stop:
        yield r
        r += step


def spline(xs, ys):
    size = len(xs)
    h = [0] + [xs[i] - xs[i - 1] for i in range(1, size)]
    dif1 = [None] + [(ys[i] - ys[i - 1]) / h[i] for i in range(1, size)]
    dif2 = ([0] +
            [(dif1[i + 1] - dif1[i]) / (xs[i + 1] - xs[i - 1])
             for i in range(1, size - 1)] +
            [0])

    xmin = min(xs)
    xmax = max(xs)

    xsort = sorted(xs)
    i = 1
    ret = []
    for x in drange(xmin, xmax, 0.01):
        while x > xsort[i]:
            i += 1

        yy0 = (dif2[i - 1] / (6 * h[i])) * ((xs[i] - x) ** 3)  # 第一項
        yy1 = (dif2[i] / (6 * h[i])) * ((x - xs[i - 1]) ** 3)  # 第２項
        yy2 = (ys[i - 1] / h[i] - h[i] * dif2[i - 1] / 6) * (xs[i] - x)  # 第３項
        yy3 = (ys[i] / h[i] - h[i] * dif2[i] / 6) * (x - xs[i - 1])  # 第４項
        y = yy0 + yy1 + yy2 + yy3
        ret.append((x, y))
    return ret


data = [
    (236.5, 186.68884),
    (237.7678, 212.64558),
    (238.76683, 224.62561),
    (239.76585, 232.6123),
    (241.7639, 239.60066),
    (242.76292, 243.59401),
    (245.76, 252.57903),
    (247.75806, 258.56903),
    (251.75415, 263.56073),
    (255.75024, 267.55408),
    (258.7522, 265.5574),
    ]

xs = [d[0] for d in data]
ys = [d[1] for d in data]


def run():
    ret = spline(xs, ys)
    for i in ret:
        print i

File: spline/spline.js
window.onload = function(){
    run();
}

function spline(xs, ys){
    var size = xs.length;
    var h = [0];
    for(var i=1; i<size; i++){
        h = h.concat(xs[i] - xs[i - 1]);
    }
    var dif1 = [null];
    for(var i=1; i<size; i++){
        dif1 = dif1.concat((ys[i] - ys[i - 1]) / h[i]);
    }
    var dif2 = [0];
    for(var i=1; i<size-1; i++){
        dif2 = dif2.concat((dif1[i + 1] - dif1[i]) / (xs[i + 1] - xs[i - 1]));
    }
    dif2 = dif2.concat(0);

    /*for(var i=0; i<dif2.length; i++){
        console.log(dif2[i]);
    }*/

    var xmin = Math.min.apply(null, xs);
    var xmax = Math.max.apply(null, xs);

    var xsort = xs.sort();
    var i = 1;
    var ret = [];
    for(var x=xmin; x<xmax; x+=0.1){
        while(x > xsort[i]){
            i++;
        }

        var yy0 = (dif2[i - 1] / (6 * h[i])) * Math.pow((xs[i] - x), 3);  // 第一項
        var yy1 = (dif2[i] / (6 * h[i])) * Math.pow((x - xs[i - 1]), 3);  // 第２項
        var yy2 = (ys[i - 1] / h[i] - h[i] * dif2[i - 1] / 6) * (xs[i] - x);  // 第３項
        var yy3 = (ys[i] / h[i] - h[i] * dif2[i] / 6) * (x - xs[i - 1]);  // 第４項
        var y = yy0 + yy1 + yy2 + yy3;
        /*if(isNaN(y)){
            console.log('-----------------');
            console.log(yy0);
            console.log(yy1);
            console.log(yy2);
            console.log(yy3);
        }*/
        ret = ret.concat([[x, y]]);
    }
    return ret;
}


function run(){
    var data = [
        [0,100],
        [100,200],
        [200,300],
        [300,350],
        [400,400],
        [500,600],
    ]

    var xs = [];
    for(var i=0; i<data.length; i++){
        xs = xs.concat(data[i][0]);
    }
    var ys = [];
    for(var i=0; i<data.length; i++){
        ys = ys.concat(data[i][1]);
    }
    var ret = spline(xs, ys);
    for(var i=0; i<ret.length; i++){
        console.log(ret[i][0] + ", " + ret[i][1]);
    }
}